packages/backend/server/src/plugins/copilot/context/resolver.ts

```diff
@@ -0,0 +1,446 @@
+    private readonly permissions: PermissionService,
+    private readonly chatSession: ChatSessionService,
+    private readonly context: CopilotContextService
+  ) {}
+
+  private async checkChatSession(
+    user: CurrentUser,
+    sessionId: string,
+    workspaceId?: string
+  ): Promise<void> {
+    const session = await this.chatSession.get(sessionId);
+    if (
+      !session ||
+      session.config.workspaceId !== workspaceId ||
+      session.config.userId !== user.id
+    ) {
+      throw new CopilotSessionNotFound();
+    }
+  }
+
+  @ResolveField(() => [CopilotContextType], {
+    description: 'Get the context list of a session',
+    complexity: 2,
+  })
+  @CallMetric('ai', 'context_create')
+  async contexts(
+    @Parent() copilot: CopilotType,
+    @CurrentUser() user: CurrentUser,
+    @Args('sessionId') sessionId: string
+  ) {
+    const lockFlag = `${COPILOT_LOCKER}:context:${sessionId}`;
+    await using lock = await this.mutex.acquire(lockFlag);
+    if (!lock) {
+      return new TooManyRequest('Server is busy');
+    }
+    await this.checkChatSession(user, sessionId, copilot.workspaceId);
+
+    return await this.context.list(sessionId);
+  }
+
+  @Mutation(() => String, {
+    description: 'Create a context session',
+  })
+  @CallMetric('ai', 'context_create')
+  async createCopilotContext(
+    @CurrentUser() user: CurrentUser,
+    @Args('workspaceId') workspaceId: string,
+    @Args('sessionId') sessionId: string
+  ) {
+    const lockFlag = `${COPILOT_LOCKER}:context:${sessionId}`;
+    await using lock = await this.mutex.acquire(lockFlag);
+    if (!lock) {
+      return new TooManyRequest('Server is busy');
+    }
+    await this.checkChatSession(user, sessionId, workspaceId);
+
+    const context = await this.context.create(sessionId);
+    return context.id;
+  }
+
+  @Mutation(() => Boolean, {
+    description: 'queue workspace doc embedding',
+  })
+  @CallMetric('ai', 'context_queue_workspace_doc')
+  async queueWorkspaceEmbedding(
+    @CurrentUser() user: CurrentUser,
+    @Args('workspaceId') workspaceId: string,
+    @Args('docId', { type: () => [String] }) docIds: string[]
+  ) {
+    await this.permissions.checkCloudWorkspace(workspaceId, user.id);
+    for (const docId of docIds) {
+      this.event.emit('workspace.doc.embedding', { workspaceId, docId });
+    }
+    return true;
+  }
+
+  @Query(() => ContextWorkspaceEmbeddingStatus, {
+    description: 'query workspace embedding status',
+  })
+  @CallMetric('ai', 'context_query_workspace_embedding_status')
+  async queryWorkspaceEmbeddingStatus(
+    @CurrentUser() user: CurrentUser,
+    @Args('workspaceId') workspaceId: string
+  ) {
+    await this.permissions.checkCloudWorkspace(workspaceId, user.id);
+
+    const total = await this.db.snapshot.count({ where: { workspaceId } });
+    const embedded = await this.db.snapshot.count({
+      where: { workspaceId, embedding: { isNot: null } },
+    });
+    return { total, embedded };
+  }
```

Comment:
It would be better to separate the workspace embedding feature into a separate PR.